src/v/storage/index_state.h

```diff
@@ -138,6 +140,15 @@ struct index_state
     // the index.
     std::optional<model::timestamp> broker_timestamp{std::nullopt};
 
+    // The offset of the first data record in the segment. This may not
+    // necessarily correspond to one of the values in the indexing vectors.
+    //
+    // Special values:
+    // - model::offset{} indicates there are no data records.
+    // - -1: indicates that the index state was written in a version that does
+    //       not include this field.
+    model::offset first_data_offset{};
```

Comment:
I wholly acknowledge how ugly this is, but I couldn't think of better way than adding this stuff to the `index_state`. 

Re: the -1, I want the semantics of a tristate, but opted to use a raw `model::offset{}` since this is written to disk. Happy to switch back if folks think the clarity upon upgrade of using a tristate is worth the extra bytes

Reply:
Sorry to waffle, I swapped this back to a tristate. This is a per-segment field so hopefully the extra bytes won't be too concerning